samples/boards/sensortile_box_pro/ble-fw-upgrade/src/main.c

```diff
@@ -0,0 +1,378 @@
+		return -1;
+	}
+	gpio_pin_configure_dt(&green_gpio, GPIO_OUTPUT_INACTIVE);
+
+	/* led 1 */
+	if (!gpio_is_ready_dt(&red_gpio)) {
+		LOG_ERR("%s: device not ready.\n", red_gpio.port->name);
+		return -1;
+	}
+	gpio_pin_configure_dt(&red_gpio, GPIO_OUTPUT_INACTIVE);
+
+	/* led 3 */
+	if (!gpio_is_ready_dt(&blue_gpio)) {
+		LOG_ERR("%s: device not ready.\n", blue_gpio.port->name);
+		return -1;
+	}
+	gpio_pin_configure_dt(&blue_gpio, GPIO_OUTPUT_INACTIVE);
+
+	return 0;
+}
+
+#define BLE_UART_BOOTLOADER_ACK  0x79
+#define BLE_UART_BOOTLOADER_NACK 0x1F
+
+#define BLE_UART_BOOTLOADER_AUTO_BAUDRATE 0x7F
+#define BLE_BL_CMD_GET_LIST               0x00
+#define BLE_BL_CMD_GET_VERSION            0x01
+#define BLE_BL_CMD_WRITE_MEMORY           0x31
+#define BLE_BL_CMD_ERASE                  0x43
+
+static uint8_t checksum(uint8_t *buffer, uint16_t len)
+{
+	uint8_t checksum = 0;
+	uint16_t i;
+
+	if (len > 1) {
+		for (i = 0; i < len; i++) {
+			checksum = checksum ^ buffer[i];
+		}
+	} else {
+		checksum = ~buffer[0];
+	}
+
+	return checksum;
+}
+
+static int ble_uart_check_ack(const struct device *bl_uart)
+{
+	int ret;
+	uint8_t ack;
+
+	while ((ret = uart_poll_in(bl_uart, &ack)) == -1) {
+	}
+
+	if (ret != 0 || ack != BLE_UART_BOOTLOADER_ACK) {
+		LOG_ERR("ble_uart_send_data: NAK (%02x)\n", ack);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int ble_uart_send_data(const struct device *bl_uart, uint8_t *data, uint16_t len)
+{
+	uint16_t i;
+
+	for (i = 0; i < len; i++) {
+		uart_poll_out(bl_uart, data[i]);
+	}
+
+	uart_poll_out(bl_uart, checksum(data, len));
+
+	return ble_uart_check_ack(bl_uart);
+}
+
+static int ble_uart_send_cmd(const struct device *bl_uart, uint8_t cmd)
+{
+	return ble_uart_send_data(bl_uart, &cmd, 1);
+}
+
+static int ble_uart_bootloader_activate(const struct device *bl_uart)
+{
+	const struct gpio_dt_spec bt_rst = GPIO_DT_SPEC_GET(DT_NODELABEL(hci_spi), reset_gpios);
+	const struct gpio_dt_spec bt_boot = GPIO_DT_SPEC_GET(DT_NODELABEL(hci_spi), irq_gpios);
+
+	/* Configure RST pin and hold BLE in Reset */
+	gpio_pin_configure_dt(&bt_rst, GPIO_OUTPUT_ACTIVE);
+	gpio_pin_configure_dt(&bt_boot, GPIO_OUTPUT_ACTIVE);
+
+	k_sleep(K_MSEC(1));
+
+	/* Take BLE out of reset */
+	gpio_pin_set_dt(&bt_rst, 0);
+	k_sleep(K_MSEC(1));
+
+	/* Send bootloader activate command */
+	uart_poll_out(bl_uart, BLE_UART_BOOTLOADER_AUTO_BAUDRATE);
+
+	return ble_uart_check_ack(bl_uart);
+}
+
+static int ble_bl_cmd_get_version(const struct device *bl_uart, uint8_t *fw_ver)
+{
+	uint8_t version = 0, opt[2];
+
+	if (ble_uart_send_cmd(bl_uart, BLE_BL_CMD_GET_VERSION) < 0) {
+		return -1;
+	}
+
+	while (uart_poll_in(bl_uart, &version) == -1) {
+	}
+
+	while (uart_poll_in(bl_uart, &opt[0]) == -1) {
+	}
+
+	while (uart_poll_in(bl_uart, &opt[1]) == -1) {
+	}
+
+	*fw_ver = version;
+
+	return ble_uart_check_ack(bl_uart);
+}
+
+static int ble_bl_cmd_mass_erase(const struct device *bl_uart)
+{
+	if (ble_uart_send_cmd(bl_uart, BLE_BL_CMD_ERASE) < 0) {
+		return -1;
```

Comment:
how about having a temporary return variable, logging the error message and returning the actual code? same in the other cases, not much point keeping it ultra compact, if something fails it'll be a pain to troubleshoot